File: flutter_base_05/lib/core/managers/websockets/ws_event_manager.dart
import 'dart:async';
import 'dart:convert';
import 'package:flutter/material.dart';
import '../../../tools/logging/logger.dart';
import 'websocket_events.dart';
import 'websocket_manager.dart';
import '../state_manager.dart';
import 'websocket_state_validator.dart';

const bool LOGGING_SWITCH = false; // Enabled for debugging WebSocket event management

/// WebSocket Event Manager - Centralized event handling for WebSocket operations
class WSEventManager {
  static final Logger _logger = Logger();
  static final WSEventManager _instance = WSEventManager._internal();
  
  // WebSocket manager instance
  final WebSocketManager _websocketManager = WebSocketManager.instance;
  
  // Event streams for different event types
  final StreamController<RoomEvent> _roomEventController = StreamController<RoomEvent>.broadcast();
  final StreamController<MessageEvent> _messageEventController = StreamController<MessageEvent>.broadcast();
  final StreamController<ConnectionStatusEvent> _connectionEventController = StreamController<ConnectionStatusEvent>.broadcast();
  final StreamController<SessionDataEvent> _sessionEventController = StreamController<SessionDataEvent>.broadcast();
  final StreamController<ErrorEvent> _errorEventController = StreamController<ErrorEvent>.broadcast();
  final StreamController<CustomEvent> _customEventController = StreamController<CustomEvent>.broadcast();
  
  // Event handlers map
  final Map<String, Function(Map<String, dynamic>)> _eventHandlers = {};
  
  // State tracking
  String? _currentRoomId;
  Map<String, dynamic>? _currentRoomInfo;
  Map<String, dynamic>? _sessionData;
  bool _isConnected = false;
  
  // Event callbacks
  final Map<String, List<Function(Map<String, dynamic>)>> _eventCallbacks = {};
  
  factory WSEventManager() => _instance;
  WSEventManager._internal();

  // Static getter for easy access
  static WSEventManager get instance {
    return _instance;
  }

  /// Initialize the event manager
  void initialize() {
    if (LOGGING_SWITCH) {
      _logger.info('🔄 Initializing WebSocket Event Manager');
    }
    _setupEventListeners();
    _registerWithStateManager();
    if (LOGGING_SWITCH) {
      _logger.info('✅ WebSocket Event Manager initialized successfully');
    }
  }

  /// Register with StateManager
  void _registerWithStateManager() {
    if (LOGGING_SWITCH) {
      _logger.debug('🔄 Registering WebSocket state with StateManager');
    }
    final stateManager = StateManager();
    stateManager.registerModuleState("websocket", {
      "isConnected": _isConnected,
      "currentRoomId": _currentRoomId,
      "currentRoomInfo": _currentRoomInfo,
      "sessionData": _sessionData,
    });
    if (LOGGING_SWITCH) {
      _logger.debug('✅ WebSocket state registered with StateManager');
    }
  }

  /// Set up event listeners for all WebSocket events
  void _setupEventListeners() {
    // Listen to WebSocket manager events
    _websocketManager.events.listen((event) {
      _handleIncomingEvent(event);
    });

    _websocketManager.connectionStatus.listen((event) {
      _handleConnectionEvent(event);
    });

    _websocketManager.messages.listen((event) {
      _handleMessageEvent(event);
    });

    _websocketManager.roomEvents.listen((event) {
      _handleRoomEvent(event);
    });

    _websocketManager.errors.listen((event) {
      _handleErrorEvent(event);
    });
  }

  /// Handle incoming WebSocket events
  void _handleIncomingEvent(WebSocketEvent event) {
    switch (event.runtimeType) {
      case ConnectionStatusEvent:
        _handleConnectionEvent(event as ConnectionStatusEvent);
        break;
      case SessionDataEvent:
        _handleSessionEvent(event as SessionDataEvent);
        break;
      case RoomEvent:
        _handleRoomEvent(event as RoomEvent);
        break;
      case MessageEvent:
        _handleMessageEvent(event as MessageEvent);
        break;
      case ErrorEvent:
        _handleErrorEvent(event as ErrorEvent);
        break;
      case CustomEvent:
        _handleCustomEvent(event as CustomEvent);
        break;
      default:
        // Unknown event type
        break;
    }
  }

  /// Handle connection status events
  void _handleConnectionEvent(ConnectionStatusEvent event) {
    _isConnected = event.status == ConnectionStatus.connected;
    _connectionEventController.add(event);
    
    // Update StateManager
    _updateStateManager();
    
    // Trigger connection callbacks
    _triggerCallbacks('connection', {
      'status': event.status.name,
      'sessionId': event.sessionId,
      'error': event.error,
    });
  }

  /// Handle session data events
  void _handleSessionEvent(SessionDataEvent event) {
    _sessionData = event.sessionData;
    _sessionEventController.add(event);
    
    // Update StateManager
    _updateStateManager();
    
    // Trigger session callbacks
    _triggerCallbacks('session', event.sessionData);
  }

  /// Handle room events
  void _handleRoomEvent(RoomEvent event) {
    switch (event.action) {
      case 'joined':
        _currentRoomId = event.roomId;
        _currentRoomInfo = event.roomData;
        break;
      case 'left':
        if (_currentRoomId == event.roomId) {
          _currentRoomId = null;
          _currentRoomInfo = null;
        }
        break;
      case 'created':
        // After room creation, the user is automatically joined
        // The room_joined event will handle setting current room
        break;
      default:
        // Unknown room action
        break;
    }
    
    // Update StateManager
    _updateStateManager();
    
    _roomEventController.add(event);
    
    // Trigger room callbacks
    _triggerCallbacks('room', {
      'action': event.action,
      'roomId': event.roomId,
      'roomData': event.roomData,
    });
  }

  /// Handle message events
  void _handleMessageEvent(MessageEvent event) {
    _messageEventController.add(event);
    
    // Trigger message callbacks
    _triggerCallbacks('message', {
      'roomId': event.roomId,
      'message': event.message,
      'sender': event.sender,
      'additionalData': event.additionalData,
    });
  }

  /// Handle error events
  void _handleErrorEvent(ErrorEvent event) {
    _errorEventController.add(event);
    
    // Trigger error callbacks
    _triggerCallbacks('error', {
      'error': event.error,
      'details': event.details,
    });
  }

  /// Handle custom events
  void _handleCustomEvent(CustomEvent event) {
    _customEventController.add(event);
    
    // Trigger custom event callbacks
    _triggerCallbacks(event.eventName, event.data);
  }

  /// Trigger callbacks for a specific event type
  void _triggerCallbacks(String eventType, Map<String, dynamic> data) {
    final callbacks = _eventCallbacks[eventType];
    if (callbacks != null) {
      for (final callback in callbacks) {
        try {
          callback(data);
        } catch (e) {
          // Error in event callback
        }
      }
    }
  }

  // ==================== PUBLIC API ====================

  /// Register a callback for a specific event type
  void onEvent(String eventType, Function(Map<String, dynamic>) callback) {
    if (!_eventCallbacks.containsKey(eventType)) {
      _eventCallbacks[eventType] = [];
    }
    _eventCallbacks[eventType]!.add(callback);
  }

  /// Unregister a callback for a specific event type
  void offEvent(String eventType, Function(Map<String, dynamic>) callback) {
    final callbacks = _eventCallbacks[eventType];
    if (callbacks != null) {
      callbacks.remove(callback);
    }
  }

  /// Register a one-time callback for a specific event type
  void onceEvent(String eventType, Function(Map<String, dynamic>) callback) {
    late Function(Map<String, dynamic>) wrappedCallback;
    wrappedCallback = (data) {
      callback(data);
      offEvent(eventType, wrappedCallback);
    };
    onEvent(eventType, wrappedCallback);
  }

  /// Trigger callbacks for a specific event type (public method)
  void triggerCallbacks(String eventType, Map<String, dynamic> data) {
    _triggerCallbacks(eventType, data);
  }

  // ==================== ROOM MANAGEMENT ====================

  /// Create a room
  Future<Map<String, dynamic>> createRoom(String userId, [Map<String, dynamic>? roomData]) async {
    try {
      final result = await _websocketManager.createRoom(userId, roomData);
      
      if (result['success'] != null) {
        // The server will automatically join the user to the room
        // and send a 'room_joined' event, which will update our state
        return result;
      } else {
        return result;
      }
    } catch (e) {
      return {'error': 'Failed to create room: $e'};
    }
  }

  /// Join a room
  Future<Map<String, dynamic>> joinRoom(String roomId, String userId) async {
    try {
      final result = await _websocketManager.joinRoom(roomId, userId);
      
      if (result['success'] != null) {
        return result;
      } else {
        return result;
      }
    } catch (e) {
      return {'error': 'Failed to join room: $e'};
    }
  }

  /// Leave a room
  Future<Map<String, dynamic>> leaveRoom(String roomId) async {
    try {
      final result = await _websocketManager.leaveRoom(roomId);
      
      if (result['pending'] != null) {
        return result;
      } else if (result['success'] != null) {
        return result;
      } else {
        return result;
      }
    } catch (e) {
      return {'error': 'Failed to leave room: $e'};
    }
  }

  /// Send a message to a room
  Future<Map<String, dynamic>> sendMessage(String roomId, String message) async {
    try {
      final result = await _websocketManager.sendMessage(roomId, message);
      
      if (result['success'] != null) {
        return result;
      } else {
        return result;
      }
    } catch (e) {
      return {'error': 'Failed to send message: $e'};
    }
  }

  /// Broadcast a message to all rooms
  Future<Map<String, dynamic>> broadcastMessage(String message) async {
    try {
      final result = await _websocketManager.broadcastMessage(message);
      
      if (result['success'] != null) {
        return result;
      } else {
        return result;
      }
    } catch (e) {
      return {'error': 'Failed to broadcast message: $e'};
    }
  }

  // ==================== STATE GETTERS ====================

  /// Get current room ID
  String? get currentRoomId => _currentRoomId;

  /// Get current room info
  Map<String, dynamic>? get currentRoomInfo => _currentRoomInfo;

  /// Get session data
  Map<String, dynamic>? get sessionData => _sessionData;

  /// Check if connected
  bool get isConnected => _isConnected;

  /// Check if user is in a specific room
  bool isInRoom(String roomId) => _currentRoomId == roomId;

  // ==================== EVENT STREAMS ====================

  /// Stream of room events
  Stream<RoomEvent> get roomEvents => _roomEventController.stream;

  /// Stream of message events
  Stream<MessageEvent> get messageEvents => _messageEventController.stream;

  /// Stream of connection events
  Stream<ConnectionStatusEvent> get connectionEvents => _connectionEventController.stream;

  /// Stream of session events
  Stream<SessionDataEvent> get sessionEvents => _sessionEventController.stream;

  /// Stream of error events
  Stream<ErrorEvent> get errorEvents => _errorEventController.stream;

  /// Stream of custom events
  Stream<CustomEvent> get customEvents => _customEventController.stream;

  // ==================== UTILITY METHODS ====================

  /// Update state manager with current state (using validated system)
  void _updateStateManager() {
    try {
      WebSocketStateUpdater.updateState({
        "isConnected": _isConnected,
        "currentRoomId": _currentRoomId,
        "currentRoomInfo": _currentRoomInfo,
        "sessionData": _sessionData,
      });
    } catch (e) {
      // Fallback to direct state manager update
      final stateManager = StateManager();
      stateManager.updateModuleState("websocket", {
        "isConnected": _isConnected,
        "currentRoomId": _currentRoomId,
        "currentRoomInfo": _currentRoomInfo,
        "sessionData": _sessionData,
      });
    }
  }

  /// Clear all event callbacks
  void clearCallbacks() {
    _eventCallbacks.clear();
  }

  /// Dispose of the event manager
  void dispose() {
    _roomEventController.close();
    _messageEventController.close();
    _connectionEventController.close();
    _sessionEventController.close();
    _errorEventController.close();
    _customEventController.close();
    clearCallbacks();
  }
} 

File: python_base_04/core/managers/websockets/ws_event_manager.py
from typing import Dict, Any, List, Callable, Optional
from datetime import datetime
import json
from enum import Enum

class EventType(Enum):
    """WebSocket event types."""
    CONNECTION = "connection"
    ROOM = "room"
    MESSAGE = "message"
    SESSION = "session"
    ERROR = "error"
    CUSTOM = "custom"

class ConnectionStatus(Enum):
    """Connection status types."""
    CONNECTING = "connecting"
    CONNECTED = "connected"
    DISCONNECTED = "disconnected"
    ERROR = "error"
    RECONNECTING = "reconnecting"

class WSEventManager:
    """WebSocket Event Manager - Centralized event handling for WebSocket operations."""
    
    def __init__(self):
        self._event_handlers: Dict[str, List[Callable]] = {}
        self._one_time_handlers: Dict[str, List[Callable]] = {}
        self._state = {
            'is_connected': False,
            'current_room_id': None,
            'current_room_info': None,
            'session_data': None,
            'connection_status': ConnectionStatus.DISCONNECTED
        }
        def register_handler(self, event_type: str, handler: Callable) -> None:
        """Register a handler for a specific event type."""
        if event_type not in self._event_handlers:
            self._event_handlers[event_type] = []
        self._event_handlers[event_type].append(handler)
        def unregister_handler(self, event_type: str, handler: Callable) -> None:
        """Unregister a handler for a specific event type."""
        if event_type in self._event_handlers:
            if handler in self._event_handlers[event_type]:
                self._event_handlers[event_type].remove(handler)
                def register_one_time_handler(self, event_type: str, handler: Callable) -> None:
        """Register a one-time handler for a specific event type."""
        if event_type not in self._one_time_handlers:
            self._one_time_handlers[event_type] = []
        self._one_time_handlers[event_type].append(handler)
        def emit_event(self, event_type: str, data: Dict[str, Any]) -> None:
        """Emit an event to all registered handlers."""
        try:
            # Add timestamp to event data
            event_data = {
                'type': event_type,
                'data': data,
                'timestamp': datetime.now().isoformat()
            }
            
            # Call regular handlers
            if event_type in self._event_handlers:
                for handler in self._event_handlers[event_type]:
                    try:
                        handler(event_data)
                    except Exception as e:
                        }")
            
            # Call one-time handlers and remove them
            if event_type in self._one_time_handlers:
                handlers_to_remove = []
                for handler in self._one_time_handlers[event_type]:
                    try:
                        handler(event_data)
                        handlers_to_remove.append(handler)
                    except Exception as e:
                        }")
                        handlers_to_remove.append(handler)
                
                # Remove called handlers
                for handler in handlers_to_remove:
                    self._one_time_handlers[event_type].remove(handler)
            
            # Update state based on event type
            self._update_state(event_type, data)
            
        except Exception as e:
            }")

    def _update_state(self, event_type: str, data: Dict[str, Any]) -> None:
        """Update internal state based on event type."""
        try:
            if event_type == EventType.CONNECTION.value:
                status = data.get('status')
                if status:
                    self._state['connection_status'] = ConnectionStatus(status)
                    self._state['is_connected'] = (status == ConnectionStatus.CONNECTED.value)
                    
            elif event_type == EventType.ROOM.value:
                action = data.get('action')
                room_id = data.get('room_id')
                room_data = data.get('room_data')
                
                if action == 'joined':
                    self._state['current_room_id'] = room_id
                    self._state['current_room_info'] = room_data
                elif action == 'left':
                    if self._state['current_room_id'] == room_id:
                        self._state['current_room_id'] = None
                        self._state['current_room_info'] = None
                        
            elif event_type == EventType.SESSION.value:
                self._state['session_data'] = data
                
        except Exception as e:
            }")

    def get_state(self) -> Dict[str, Any]:
        """Get current state."""
        return self._state.copy()

    def get_connection_status(self) -> ConnectionStatus:
        """Get current connection status."""
        return self._state['connection_status']

    def is_connected(self) -> bool:
        """Check if currently connected."""
        return self._state['is_connected']

    def get_current_room_id(self) -> Optional[str]:
        """Get current room ID."""
        return self._state['current_room_id']

    def get_current_room_info(self) -> Optional[Dict[str, Any]]:
        """Get current room info."""
        return self._state['current_room_info']

    def get_session_data(self) -> Optional[Dict[str, Any]]:
        """Get current session data."""
        return self._state['session_data']

    def clear_handlers(self, event_type: Optional[str] = None) -> None:
        """Clear all handlers or handlers for a specific event type."""
        if event_type:
            if event_type in self._event_handlers:
                del self._event_handlers[event_type]
            if event_type in self._one_time_handlers:
                del self._one_time_handlers[event_type]
            else:
            self._event_handlers.clear()
            self._one_time_handlers.clear()
            def get_handler_count(self, event_type: str) -> int:
        """Get the number of handlers for a specific event type."""
        regular_count = len(self._event_handlers.get(event_type, []))
        one_time_count = len(self._one_time_handlers.get(event_type, []))
        return regular_count + one_time_count

    def list_registered_events(self) -> List[str]:
        """List all registered event types."""
        events = set()
        events.update(self._event_handlers.keys())
        events.update(self._one_time_handlers.keys())
        return list(events)

    def handle_connection_event(self, status: str, session_id: Optional[str] = None, error: Optional[str] = None) -> None:
        """Handle connection status events."""
        event_data = {
            'status': status,
            'session_id': session_id,
            'error': error
        }
        self.emit_event(EventType.CONNECTION.value, event_data)

    def handle_room_event(self, action: str, room_id: str, room_data: Optional[Dict[str, Any]] = None) -> None:
        """Handle room events."""
        event_data = {
            'action': action,
            'room_id': room_id,
            'room_data': room_data or {}
        }
        self.emit_event(EventType.ROOM.value, event_data)

    def handle_message_event(self, room_id: str, message: str, sender: str, additional_data: Optional[Dict[str, Any]] = None) -> None:
        """Handle message events."""
        event_data = {
            'room_id': room_id,
            'message': message,
            'sender': sender,
            'additional_data': additional_data or {}
        }
        self.emit_event(EventType.MESSAGE.value, event_data)

    def handle_session_event(self, session_data: Dict[str, Any]) -> None:
        """Handle session events."""
        self.emit_event(EventType.SESSION.value, session_data)

    def handle_error_event(self, error: str, details: Optional[str] = None) -> None:
        """Handle error events."""
        event_data = {
            'error': error,
            'details': details
        }
        self.emit_event(EventType.ERROR.value, event_data)

    def handle_custom_event(self, event_name: str, data: Dict[str, Any]) -> None:
        """Handle custom events."""
        event_data = {
            'event_name': event_name,
            'data': data
        }
        self.emit_event(event_name, event_data)

    def create_room_event(self, room_id: str, room_data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a room event data structure."""
        return {
            'action': 'created',
            'room_id': room_id,
            'room_data': room_data,
            'timestamp': datetime.now().isoformat()
        }

    def create_join_room_event(self, room_id: str, room_data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a join room event data structure."""
        return {
            'action': 'joined',
            'room_id': room_id,
            'room_data': room_data,
            'timestamp': datetime.now().isoformat()
        }

    def create_leave_room_event(self, room_id: str) -> Dict[str, Any]:
        """Create a leave room event data structure."""
        return {
            'action': 'left',
            'room_id': room_id,
            'timestamp': datetime.now().isoformat()
        }

    def create_message_event(self, room_id: str, message: str, sender: str, additional_data: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Create a message event data structure."""
        return {
            'room_id': room_id,
            'message': message,
            'sender': sender,
            'additional_data': additional_data or {},
            'timestamp': datetime.now().isoformat()
        }

    def create_connection_event(self, status: str, session_id: Optional[str] = None, error: Optional[str] = None) -> Dict[str, Any]:
        """Create a connection event data structure."""
        return {
            'status': status,
            'session_id': session_id,
            'error': error,
            'timestamp': datetime.now().isoformat()
        }

    def create_error_event(self, error: str, details: Optional[str] = None) -> Dict[str, Any]:
        """Create an error event data structure."""
        return {
            'error': error,
            'details': details,
            'timestamp': datetime.now().isoformat()
        }

    def create_custom_event(self, event_name: str, data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a custom event data structure."""
        return {
            'event_name': event_name,
            'data': data,
            'timestamp': datetime.now().isoformat()
        }

    def validate_event_data(self, event_type: str, data: Dict[str, Any]) -> bool:
        """Validate event data structure."""
        try:
            if event_type == EventType.CONNECTION.value:
                return 'status' in data
            elif event_type == EventType.ROOM.value:
                return 'action' in data and 'room_id' in data
            elif event_type == EventType.MESSAGE.value:
                return 'room_id' in data and 'message' in data and 'sender' in data
            elif event_type == EventType.SESSION.value:
                return isinstance(data, dict)
            elif event_type == EventType.ERROR.value:
                return 'error' in data
            else:
                # Custom event
                return isinstance(data, dict)
        except Exception as e:
            }")
            return False

    def get_event_statistics(self) -> Dict[str, Any]:
        """Get statistics about registered events."""
        stats = {
            'total_events': len(self.list_registered_events()),
            'regular_handlers': {},
            'one_time_handlers': {},
            'state': self._state
        }
        
        for event_type in self._event_handlers:
            stats['regular_handlers'][event_type] = len(self._event_handlers[event_type])
            
        for event_type in self._one_time_handlers:
            stats['one_time_handlers'][event_type] = len(self._one_time_handlers[event_type])
            
        return stats

    def reset_state(self) -> None:
        """Reset the internal state."""
        self._state = {
            'is_connected': False,
            'current_room_id': None,
            'current_room_info': None,
            'session_data': None,
            'connection_status': ConnectionStatus.DISCONNECTED
        }
        def cleanup(self) -> None:
        """Clean up the event manager."""
        self.clear_handlers()
        self.reset_state()
        